### DISCUSSION TITLE
Nested Collections (Taxonomies) (👍84)

### DISCUSSION THREAD

**Discussion by zachleat** [11ty/eleventy#3991] [Enhancement Queue]
I believe some other tools call this `taxonomies`

This could enable some other neat things like nested levels of pagination. An example of a workaround for two levels of pagination can be found here https://github.com/11ty/eleventy/issues/332

Sold best by this tweet: https://twitter.com/sirinath/status/1230855052547514369

> Tags can be hierarchically bundled together by different criteria. E.g. tag 'rust' in category 'programming' can be kept sperate from tag 'rust' in 'metallurgy'. 'Star' in 'music' can be separated from 'star' in 'cosmology' in a 'series' of posts on 'raggae'.

See also

* Hugo https://gohugo.io/content-management/taxonomies/
* WordPress https://developer.wordpress.org/themes/basics/categories-tags-custom-taxonomies/

**Comment by zachleat** [comment]
> This repository is now using [lodash style issue management](https://twitter.com/samselikoff/status/991395669016436736) for enhancements. This means enhancement issues will now be closed instead of leaving them open.

> View [the enhancement backlog here](https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+). Don’t forget to upvote the top comment with 👍!

**Comment by sirinath** [comment]
Best writeup on this is:

- [5 steps to organizing your blog’s content—taxonomy simplified](https://www.thisisvisceral.com/news/5-steps-organizing-blog-content/) - scroll down for the content to appear.

Also might be interesting:

- [What Is Taxonomy, and Why Does It Matter?](https://wordpress.com/go/solve/what-is-taxonomy-and-why-does-it-matter/)

Other tools which have it:

- Hugo [Taxonomies](https://gohugo.io/content-management/taxonomies/)
- Zola [Taxonomies](https://www.getzola.org/documentation/content/taxonomies/)
- WordPress - [Categories, Tags, & Custom Taxonomies](https://developer.wordpress.org/themes/basics/categories-tags-custom-taxonomies/)

**Comment by paulshryock** [comment]
I feel like custom taxonomies are already built into eleventy, even if it's not documented as such.

For example:

```
---
title: 'Hello World'
contentType: 'article'
category: 'news'
tags: [ 'greeting', 'friendly' ]
---
```

In `.eleventy.js`, create an `articles` collection:

```js
module.exports = function (eleventyConfig) {
  eleventyConfig.addCollection('articles', collection => {
    return collection.getAll().filter(post => post.data.contentType === 'article'))
  })
}
```

Then in a Javascript data file, query and filter data as needed.

```js
data.articles = collections.articles.filter((article) => {
  return article.category === 'news' && article.tags.includes('greeting')
})
```

Something like that?

Although I'm not sure if it's possible to access the Eleventy `collections` object from JavaScript data files.

**Comment by sielay** [comment]
I don't want here to promote my plugin, but you can find here examples how to create taxonomies and paginate on them. I'm still to make few improvements. First thing I'd like do anyhow is to allow disabling native support for tags, as it makes such approach a bit more problematic.

https://github.com/sielay/eleventy-plugin-blog/blob/master/index.js#L156

All is mutation of idea promoted by @jeromecoupe https://www.webstoemp.com/blog/basic-custom-taxonomies-with-eleventy/

**Comment by hidegh** [comment]
@zachleat I'd also consider **nested collections**, which **HEXO** has out of the box e.g.: /collections/technical/testing/1 or  /collections/technical/testing/unit-tests/1

Again here (beside nesting) we can have 2 things to solve:
1. do we want to display items (and counts) to a category (we might have just a pure nested category listing)
2. if so, do we want to display immediate items only (or do we want to include items that belong to the subcategories of the currently selected category)

Paging can be also interesting, as we could page over:
1. the merged sub-category + item list
2. only over items

**Comment by m4rrc0** [comment]
Hey everyone. After doing some digging, I think I can safely say that this feature request's importance is being overlooked. I have found many Github issues pointing to the same kind of request. Of course, some are duplicates but you can look at it any way you want, many people have this problem and different solutions are used while not exactly knowing what the impact is on build performance, maintainability, flexibility of the data structure, ...

In the [enhancement backlog](https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+) alone, nr 7 and 8 (in order of number of votes) point to the same request.

Then there are requests like [Programmatically create pages from data files #136](https://github.com/11ty/eleventy/issues/136) which in fact point to the same issue of lacking control over how pages are built and/or dissociating the relation between 1 file => 1 flat collection.

I totally get that not everyone needs this feature and that there are workarounds. It just seems to me that there are more people needing this than meets the eye.

To try and make my case, and in an effort of compiling what has been discussed on the topic, here are the issues and discussions I could find about this topic (sorry this isn't sorted in any relevant manner):
- https://github.com/11ty/eleventy/issues/332
- https://github.com/11ty/eleventy/issues/1412
- https://github.com/11ty/eleventy/pull/1559
- https://github.com/11ty/eleventy/issues/461
- https://github.com/11ty/eleventy/issues/294
- https://github.com/11ty/eleventy/issues/308
- https://github.com/11ty/eleventy/issues/427
- https://github.com/11ty/eleventy/issues/502
- https://github.com/11ty/eleventy/issues/316
- https://github.com/11ty/eleventy/issues/772
- https://github.com/11ty/eleventy/issues/964
- https://github.com/11ty/eleventy/issues/1357
- https://github.com/11ty/eleventy/issues/1284
- https://github.com/11ty/eleventy/issues/776
- https://github.com/11ty/eleventy/issues/793
- https://github.com/11ty/eleventy/issues/955
- https://github.com/11ty/eleventy/issues/1120
- https://github.com/11ty/eleventy/issues/2141
- https://github.com/11ty/eleventy/pull/1711
- https://github.com/11ty/eleventy/pull/1697
- https://github.com/11ty/eleventy/issues/699

Non core projects trying to solve this:
- https://github.com/jeromecoupe/11ty-paginate-taxonomies with [this article](https://www.webstoemp.com/blog/basic-custom-taxonomies-with-eleventy/) covering the method I believe
- https://github.com/sielay/eleventy-plugin-blog
- https://github.com/solution-loisir/eleventy-deep-pagination with [this demo website](https://eleventy-deep-pagination.netlify.app/)
- https://gist.github.com/MWDelaney/6dd60be5e4c223217d597983d20bde20
- https://github.com/MWDelaney/ZeroPoint/pull/31


I challenge everyone to find more (non-trivial) issues on any other topic! 🤣

I think all those requests have found workarounds to make their use cases work but there is still no canonical way to handle this in Eleventy I believe (with all the obvious short and long-term consequences of hacking around an issue).

Isn't it time to have a proper discussion about this? 😄

(And in the meantime I'll find my own workaround 😅)

**Comment by zachleat** [comment]
Cross linking to new home base for Double Layer Pagination https://github.com/11ty/eleventy/discussions/3987
